### PR TITLE
Unnamed fields states (substates) support in derive(States)

### LINKED ISSUE #8187: Add support for enum fields in the #[derive(States)
## What problem does this solve or what need does it fill?
For me, I'm developing a visual novel engine. I wrote a state for managing the navigation of the app. Since this engine focuses on extensibility, there needs a way for users to add their own state to work inside my state. In Rust, that's possible with type arguments and fields (`enum Foobar(T){ Foo, Bar(T), }`)

That's also how it was done in [Inox2d](https://docs.rs/inox2d/latest/inox2d/nodes/node_data/enum.InoxData.html). 

However, in the implementation of the States derive macro, [enum fields are not supported](       https://github.com/bevyengine/bevy/blob/4f16d6e0dc14c12ab29b72880d5d3d22003eed72/crates/bevy_ecs/macros/src/states.rs#LL12C15-L12C15).

## What solution would you like?
As described in issue title.

## What alternative(s) have you considered?
We can implement `States` manually, but won't that be tedious?

Or we can "flatten" the enums. So from:
```rs
pub enum Foo {
   Fieldless
   Bar(Bar)
}
pub enum Bar {
   Alice,
   Bob
}
```
to:
```rs
pub enum Foo {
 Fieldless,
 BarAlice,
 BarBob
}
```

But for adding states that we can't control (e.g. user-defined states), that won't be possible.

## Additional context
N/A

### PR DESCRIPTION
**Pull request by natasria**
# Objective
- Fixes #8187 
## Solution
TODO
This requires that all unnamed fields implement States too, obviously.
~~This does not work yet, since we have to convert Vec's IntoIter type into array's IntoIter type, which isn't possible, since we have to specify the length at runtime. Unfortunately we can't get the variants length at compile time.~~ Now works.

## Changelog
- Added support for enum fields states in #[derive(States)]

## Migration Guide
The Self::Iter type of "fieldful" enums that derive the States macro is `std::vec::IntoIter` rather than `std::array::IntoIter`, which remains to be the Iter type of fieldless enums. This is because child variations have to be collected at runtime. This does not change anything in the [State](https://docs.rs/bevy/latest/bevy/ecs/prelude/trait.States.html) trait though, so users are less likely to be affected.

However, this might affect some edge cases, so I think I will include this just so that everyone know this detail..

### CODE REVIEW

**Comment by natalia-asteria** [comment]
I was getting the same error that's displayed on the Miri test on Rust Analyzer, but I wonder what was causing it.

In my machine, it now works though.
A screenshot of VSCode displaying that this pull request has been successfully built on Rust Analyzer and Cargo CLI

File: crates/bevy_ecs/macros/src/states.rs [hunk 1]
```diff
@@ -28,15 +26,46 @@ pub fn derive_states(input: TokenStream) -> TokenStream {
...
+    let (variants_impl, iter_type) = if fieldful_idents.len() > 0 {
+        (
+            quote! {
+                let mut fields = vec![#(Self::#fieldless_idents,)*];
+                let fieldful = [#(<Self::#fieldful_idents as States>::variants().map(|variant| {
+                            Self::#fieldful_idents(variant)
+                }),)*];
+                for field in fieldful {
+                    fields.extend(field)
+                }
```
Comment (MrGVSV) [inline crates/bevy_ecs/macros/src/states.rs:51]: I feel like you could probably use [`flat_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flat_map) here and remove the extra array and for loop, right?

File: crates/bevy_ecs/macros/src/states.rs [hunk 2]
```diff
@@ -28,15 +26,46 @@ pub fn derive_states(input: TokenStream) -> TokenStream {
...
+        .iter()
+        .filter(|v| v.fields.is_empty())
+        .map(|v| &v.ident);
+    let fieldful_idents: Vec<&Ident> = enumeration
+        .variants
+        .iter()
+        .filter(|v| !v.fields.is_empty())
+        .map(|v| &v.ident)
+        .collect();
+    // kagamine
```
Comment (MrGVSV) [inline crates/bevy_ecs/macros/src/states.rs:40]:
```suggestion
```

File: crates/bevy_ecs/macros/src/states.rs (hunk 1 above)
Comment (natalia-asteria) [reply crates/bevy_ecs/macros/src/states.rs:51]: Ah, okay

File: crates/bevy_ecs/macros/src/states.rs (hunk 2 above)
Comment (natalia-asteria) [reply crates/bevy_ecs/macros/src/states.rs:40]: :eyes:

Comment (natalia-asteria) [reply crates/bevy_ecs/macros/src/states.rs:40]: I forgot to remove it

File: crates/bevy_ecs/macros/src/states.rs (hunk 1 above)
Comment (natalia-asteria) [reply crates/bevy_ecs/macros/src/states.rs]: I'm not sure if flat_map is the right option? I changed to `flatten` though.

Comment (natalia-asteria) [reply crates/bevy_ecs/macros/src/states.rs]: A screenshot of Cargo lint complaining about not using flat_map
Ah, I see what you mean.

**Comment by natalia-asteria** [comment]
[99e55fd](https://github.com/bevyengine/bevy/pull/8188/commits/99e55fd8e6c0d634d6182d875598b02122a3e3fd) should fix this macro. If anyone has issues, let me know. ^_^

File: crates/bevy_ecs/macros/src/states.rs (hunk 1 above)
Comment (natalia-asteria) [reply crates/bevy_ecs/macros/src/states.rs]: Wait, this was different though.

**Comment by natasria** [comment]
CI done.

File: crates/bevy_ecs/macros/src/states.rs [hunk 3]
```diff
@@ -9,36 +10,90 @@ pub fn derive_states(input: TokenStream) -> TokenStream {
     let error = || {
         syn::Error::new(
             Span::call_site().into(),
-            "derive(States) only supports fieldless enums",
+            "derive(States) only supports enums (additionaly, all of the fields need to implement States too)",
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/macros/src/states.rs:13]: It would be nice to be able to split apart these error cases, but I won't block on it.

**Comment by hymm** [comment]
One of my worries with supporting this is that it looks like nested states, but is missing the ability to schedule something in something like `OnEnter(MyState(_))`. i.e. schedule something when entering MyState for any substate.

**Comment by alice-i-cecile** [comment]
Agreed: that's an important bit of functionality that users would want rapidly. I'm tempted to say that we need to merge that at the same time.

**Comment by natasria** [comment]
> One of my worries with supporting this is that it looks like nested states, but is missing the ability to schedule something in something like `OnEnter(MyState(_))`. i.e. schedule something when entering MyState for any substate.

Also, considering my usecase, I wonder.
How will plugin crates interact with the state? For the core state of my project, I could use State<T> that has State::Custom(T), and application users can track the State state to get the T, but what about library users?

File: crates/bevy_ecs/macros/src/states.rs (hunk 3 above)
Comment (natasria) [reply crates/bevy_ecs/macros/src/states.rs:13]: I think I intended to write that too, but wondered how to write it in a way that isn't scary to users. 

But since you support it, great!

**Comment by natasria** [comment]
> One of my worries with supporting this is that it looks like nested states, but is missing the ability to schedule something in something like `OnEnter(MyState(_))`. i.e. schedule something when entering MyState for any substate.

I wonder how should the API be like though, 

Personally, the ideal API would be: 
```rs
app.add_enter_system(MyState::MainMenu, setup_menu);
``` 
But in Rust, unnamed enum variants have to be specified. It's possible with a macro, but do we really need macros?
Or what about implementing a function or a constant in the State that can act as a marker for the substate? So if there's the `MainMenu` variant, the `derive(State)` macro will generate MAIN_MENU or main_menu().

**Comment by natasria** [comment]
So, maybe. States can be made to implement PartialEq to the `MAIN_MENU` (or anything else) constants. 
Users can just write:
```rs
app.add_enter_system(MyState::MAIN_MENU, setup_menu);
```
and then [here](https://github.com/bevyengine/bevy/blob/4f16d6e0dc14c12ab29b72880d5d3d22003eed72/crates/bevy_ecs/src/schedule/state.rs#L125), we can check for any substates involved in this, and if there are, also run the schedules associated with them. I think that'll be fairly easy to implement. Just need consensus.

**Comment by natasria** [comment]
WAIT, just realized. You can leave unnamed fields. That'll make it into a function.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=520116fedf8330c160bfbc268db3bb4c

**Comment by natasria** [comment]
Okay, so now would be how to have functions be accepted as a schedule label.

**Comment by natasria** [comment]
> Agreed: that's an important bit of functionality that users would want rapidly. I'm tempted to say that we need to merge that at the same time.

@alice-i-cecile should I write that in a separate PR or in this PR?

**Comment by natasria** [comment]
> Also, considering my usecase, I wonder. How will plugin crates interact with the state? For the core state of my project, I could use State that has State::Custom(T), and application users can track the State state to get the T, but what about library users?

Thinking again, function signatures like `Fn(MyLibraryState)` -> T might be useful?

**Comment by natasria** [comment]
Also, I'm not sure how to implement variations() for unnamed fields that have multiple arguments. That means we have to calculate all possible combinations right?

**Comment by natasria** [comment]
I think we can use this for that.. but new dependency.. https://docs.rs/itertools/latest/itertools/macro.iproduct.html

**Comment by natasria** [comment]
> > Agreed: that's an important bit of functionality that users would want rapidly. I'm tempted to say that we need to merge that at the same time.
> 
> @alice-i-cecile should I write that in a separate PR or in this PR?

Umm... just realized. Should be in a separate PR.

**Comment by alice-i-cecile** [comment]
Yep, separate PR :) IMO this PR is blocked on that work being done, but we can split it up to ease review.
